lib/SILGen/SILGenBridging.cpp

```diff
@@ -1548,7 +1622,75 @@ void SILGenFunction::emitNativeToForeignThunk(SILDeclRef thunk) {
   SILValue nativeFn = emitGlobalFunctionRef(loc, native, nativeInfo);
 
   SILValue result;
-  assert(foreignError.hasValue() == substTy->hasErrorResult());
+  
+  // Helper function to pass a native async function's result as arguments to
+  // the ObjC completion handler block.
+  auto passResultToCompletionHandler = [&](SILValue result) -> SILValue {
+    Scope completionArgScope(*this, loc);
+    
+    SmallVector<SILValue, 2> completionHandlerArgs;
+    auto completionTy = foreignAsyncSlot->getType().castTo<SILFunctionType>();
+    
+    auto asyncResult = emitManagedRValueWithCleanup(result);
+    
+    auto pushArg = [&](ManagedValue arg,
+                       CanType nativeFormalTy,
+                       SILParameterInfo param) {
+      auto bridgedTy = param.getInterfaceType();
+      auto bridgedArg = emitNativeToBridgedValue(loc,
+                                 arg, nativeFormalTy,
+                                 bridgedTy,
+                                 SILType::getPrimitiveObjectType(bridgedTy));
+      completionHandlerArgs.push_back(bridgedArg.borrow(*this, loc).getValue());
+    };
+    
+    auto errorParamIndex = foreignAsync->completionHandlerErrorParamIndex();
+    auto pushErrorPlaceholder = [&]{
+      auto errorArgTy = completionTy->getParameters()[*errorParamIndex]
+        .getSILStorageInterfaceType();
+      
+      // Error type must be optional. We pass nil for a successful return
+      auto none = B.createOptionalNone(loc, errorArgTy);
+      completionHandlerArgs.push_back(none);
+    };
+    
+    unsigned numResults
+      = completionTy->getParameters().size() - errorParamIndex.hasValue();
+    
+    if (numResults == 1) {
+      if (errorParamIndex && *errorParamIndex == 0) {
+        pushErrorPlaceholder();
+      }
+      pushArg(asyncResult,
+              nativeFormalResultType,
+              completionTy->getParameters()[completionHandlerArgs.size()]);
+
+      if (errorParamIndex && *errorParamIndex == 1) {
+        pushErrorPlaceholder();
+      }
+    } else {
+      // A tuple return maps to multiple completion handler parameters.
+      auto formalTuple = cast<TupleType>(nativeFormalResultType);
+      
+      for (unsigned paramI : indices(completionTy->getParameters())) {
+        if (errorParamIndex && paramI == *errorParamIndex) {
+          pushErrorPlaceholder();
+          continue;
+        }
+        auto elementI = paramI - (errorParamIndex && paramI > *errorParamIndex);
+        auto param = completionTy->getParameters()[paramI];
+        auto formalTy = formalTuple.getElementType(elementI);
+        auto argPiece = B.createTupleExtract(loc, asyncResult, elementI);
+        pushArg(argPiece, formalTy, param);
+      }
+    }
+    // Pass the bridged results on to the completion handler.
+    B.createApply(loc, foreignAsyncSlot, {}, completionHandlerArgs);
+    
+    // The immediate function result is an empty tuple.
+    return SILUndef::get(SGM.Types.getEmptyTupleType(), F);
```

Comment:
I think in other places we emit a tuple instruction. Does an undef of empty tuple type have the same semantics?

Reply:
`undef` is "an arbitrary value of the type", and empty tuple only has one value, so yes.